v-next/hardhat-utils/src/bytecode.ts

```diff
@@ -0,0 +1,89 @@
+import type { PrefixedHexString } from "./hex.js";
+import type {
+  Artifact,
+  LibraryAddresses,
+  LibraryLink,
+} from "./internal/bytecode.js";
+
+import { getPrefixedHexString, getUnprefixedHexString } from "./hex.js";
+import {
+  checkAmbiguousOrUnnecessaryLinks,
+  checkMissingLibraryAddresses,
+  checkOverlappingLibraryNames,
+  checkProvidedLibraryAddresses,
+} from "./internal/bytecode.js";
+
+/**
+ * Resolves the linked bytecode for a given contract artifact by ensuring all
+ * required libraries are correctly linked.
+ *
+ * @param artifact The contract artifact containing the bytecode and link references.
+ * @param providedLibraries An object containing library names as keys and their addresses as values.
+ * @returns The linked bytecode with all required libraries correctly linked.
+ * @throws InvalidLibraryAddressError If any provided library address is invalid.
+ * @throws AmbiguousLibraryNameError If any provided library name matches multiple needed libraries.
+ * @throws UnnecessaryLibraryError If any provided library name is not needed by the contract.
+ * @throws OverlappingLibrariesError If any library is provided more than once.
+ * @throws MissingLibrariesError If any needed library address is missing.
+ */
+export function resolveLinkedBytecode(
+  artifact: Artifact,
+  providedLibraries: LibraryAddresses,
+): PrefixedHexString {
+  checkProvidedLibraryAddresses(providedLibraries);
+
+  const neededLibraries: LibraryLink[] = [];
+  for (const [sourceName, sourceLibraries] of Object.entries(
+    artifact.linkReferences,
+  )) {
+    for (const libraryName of Object.keys(sourceLibraries)) {
+      const libraryFqn = `${sourceName}:${libraryName}`;
+      const address =
+        providedLibraries[libraryFqn] ?? providedLibraries[libraryName];
+
+      neededLibraries.push({
+        sourceName,
+        libraryName,
+        libraryFqn,
+        address,
+      });
+    }
+  }
+
+  checkAmbiguousOrUnnecessaryLinks(providedLibraries, neededLibraries);
+  checkOverlappingLibraryNames(providedLibraries, neededLibraries);
+  checkMissingLibraryAddresses(neededLibraries);
+
+  return linkBytecode(artifact, neededLibraries);
+}
+
+/**
+ * Links the bytecode of a contract artifact with the provided library addresses.
+ * This function does not perform any validation on the provided libraries.
+ *
+ * @param artifact The contract artifact containing the bytecode and link references.
+ * @param libraries An array of LibraryLink objects representing the libraries to be linked.
+ * @returns The linked bytecode with all provided libraries correctly linked.
+ */
+export function linkBytecode(
```

Comment:
Are we expecting this to be used outside of `resolveLinkedBytecode`? Or is this exported for for testing?

Reply:
I don’t expect to see a lot of usage of `linkBytecode`, but the idea was to export it for cases where validation and error handling need to be handled externally.